src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,37 @@
+package mate.academy.service.impl;
+
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.AuthenticationService;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private static UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userFromDbOptional = userService.findByEmail(email);
+        if (userFromDbOptional.isPresent()
+                && HashUtil.hashPassword(password, userFromDbOptional.get().getSalt())
+                        .equals(userFromDbOptional.get().getPassword())) {
+            return userFromDbOptional.get();
+        }
+        throw new AuthenticationException("There is not such login " + email
+                + "in DB, you should register first.");
+    }
+
+    @Override
+    public User register(String email, String password) throws AuthenticationException {
+        if (userService.findByEmail(email).isPresent()) {
+            throw new AuthenticationException("This user is already registered");
```

Comment:
I think it's better to throw another exception (like RegistrationException)

Reply:
I think both (login and register) are trying to authenticate the user. If u do as u said, it will be better to create a parent class AuthenticationException and two kids RegistrationException and LoginException. But I can't understand why))

Reply:
Registration, Authentication and Authorization these are all different processes. Authentication is the process of identifying a user to provide access to a system. Registration is the process of adding a new user to the system. Therefore, it seems to me more logical to throw the registration exception when registering a user.

Reply:
gocha